SCClassLibrary/Common/Core/CondVar.sc

```diff
@@ -0,0 +1,116 @@
+CondVar {
+	var waitingThreads;
+
+	*new { ^super.newCopyArgs(Array.new, Array.new) }
+
+	wait { |predicate|
+		if(predicate.isNil) {
+			this.prWait
+		} {
+			// If the predicate is already true, we return immediately. If not,
+			// we block until the predicate is true after being signalled.
+			while { predicate.value.not } { this.prWait }
+		}
+	}
+
+	waitFor { |timeoutBeats, predicate|
+		var endingBeats;
+
+		// Convert timeoutBeats to a value we know is safe to call `wait` on,
+		// since an exception thrown in the timeout thread is a very bad thing.
+		// May throw if preconditions are not met.
+		timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);
+
+		// This is a slightly more complex version of the body of `wait`. The
+		// extra complexity comes from maintaining the timeout requirement.
+		if(predicate.isNil) {
+			// Interface requirement: immediately return if timeout is non-positive
+			if(timeoutBeats <= 0) { ^false };
+			^this.prWaitFor(timeoutBeats)
+		} {
+			endingBeats = thisThread.beats + timeoutBeats;
+
+			// If the predicate is already true, we return true immediately.
+			while {
+				predicate.value.not
+			} {
+				// We only get here when the predicate is not true, so we can return
+				// false immediately if we are timed out.
+				if(thisThread.beats >= endingBeats) { ^false };
+				// If the timeout expires, return the value of the predicate; otherwise,
+				// we were signalled, so go back through the loop.
+				if(this.prWaitFor(endingBeats - thisThread.beats).not) { ^predicate.value };
+			};
+
+			^true
+		}
+	}
+
+	signalOne {
+		if (waitingThreads.isEmpty.not) {
+			this.prWakeThread(waitingThreads.removeAt(0));
+		}
+	}
+
+	signalAll {
+		waitingThreads.do(this.prWakeThread(_));
+		waitingThreads = Array.new;
+	}
+
+	shallowCopy { this.shouldNotImplement(thisMethod) }
```

Comment:
Maybe worth noting in a comment that `copy` dispatches to Object:copy --> CondVar:shallowCopy?